proteuslib/flowsheets/RO_with_energy_recovery/RO_with_energy_recovery.py

```diff
@@ -146,14 +157,15 @@ def set_operating_conditions(m, water_recovery=0.5, over_pressure=0.3, solver=No
 
     # ---specifications---
     # feed
-    feed_flow_mass = 1  # feed mass flow rate [kg/s]
-    feed_mass_frac_NaCl = 0.035  # feed NaCl mass fraction [-]
-    feed_mass_frac_H2O = 1 - feed_mass_frac_NaCl # feed H20 mass fraction [-]
-
-    m.fs.feed.flow_mass_phase_comp[0, 'Liq', 'NaCl'].fix(feed_flow_mass * feed_mass_frac_NaCl)
-    m.fs.feed.flow_mass_phase_comp[0, 'Liq', 'H2O'].fix(feed_flow_mass * feed_mass_frac_H2O)
-    m.fs.feed.pressure.fix(101325)  # atmospheric pressure [Pa]
-    m.fs.feed.temperature.fix(273.15 + 25)  # room temperature [K]
+    # state variables
+    m.fs.feed.properties[0].pressure.fix(101325)  # feed pressure [Pa]
+    m.fs.feed.properties[0].temperature.fix(273.15 + 25)  # feed temperature [K]
+    # properties (cannot be fixed for initialization routines, must calculate the state variables)
+    m.fs.feed.properties.calculate_state(
+        var_args={('flow_vol_phase', 'Liq'): 1e-3,  # feed volumetric flow rate [m3/s]
+                  ('mass_frac_phase_comp', ('Liq', 'NaCl')): 0.035},  # feed NaCl mass fraction [-]
```

Comment:
Just to understand better--would users technically be able to enter any variables they want here? 

It would be interesting to see if @bknueven could repurpose his approach for assessing stability with parameter_sweep to this `calculate_state` method. The intention would be to probe common variable inputs and their effect on stability, which we could then advise the user about in documentation (and raise exceptions for combos we just don't want to allow).

Reply:
I wonder if this is the "right" interface. An alternative is:
```python
var_args={'flow_vol_phase[Liq]': 1e-3,  # feed volumetric flow rate [m3/s]
          'mass_frac_phase_comp[Liq, NaCl]': 0.035},  # feed NaCl mass fraction [-]
```
and using `find_component` or `component` in the `calculate_state` method implementation as oppose to `getattr`.

As for parameter_sweep, I think it would need some changes to know to look for a `calculate_state` method and know to do the appropriate thing. It's certainly possible, but I imagine it getting challenging for the user to communicate what gets set where, in additional to the effects of this method on the warm-start initialization provided by the prior solve.

Reply:
Regarding combinations we wish to exclude, I think that is relatively straightforward based on some work I did last week. We can define polyhedral regions in the unit model's variables and test for inclusion using ipopt or a linear programming solver.

Reply:
> I wonder if this is the "right" interface. An alternative is:
> 
> ```python
> var_args={'flow_vol_phase[Liq]': 1e-3,  # feed volumetric flow rate [m3/s]
>           'mass_frac_phase_comp[Liq, NaCl]': 0.035},  # feed NaCl mass fraction [-]
> ```
I think I agree with @bknueven on this one, since it seems the format he is proposing is more in line with that of IDAES and other parts of Proteuslib.

Reply:
@adam-a-a and @bknueven ok you guys have pointed out two issues: 1) how does the parameter sweep tool know to use calculate_state, 2) API for inputting variables and an attribute should be consistent. 

After thinking about it, I don't think 1 is an issue because the parameter sweep likely doesn't need to use it. This calculate_state method would primarily be used for determining state variables from properties before initialization. Since users have to initialize the model before using the parameter sweep tool, they can just fix the properties (and unfix the state variables for the appropriate degrees of freedom) in subsequent simulation/optimization.

I think 2 is a fair point, but its hard to address because the API isn't consistent across IDAES and ProteusLib. I also want to loop @andrewlee94 for this API discussion. For example there are equivalent things like the following:

1. The form used in IDAES scaling methods, ProteusLib property test harness, and this PR where the dictionary key is a tuple with the variable name string and its index:
```python
 var_dict = {('temperature', None): 300,
             ('flow_mass_phase_comp', ('Liq', 'NaCl')): 1,
             ('flow_mass_phase_comp', ('Liq', 'H2O')): 2,
            }
```
2. The form used in IDAES initialization routines where the dictionary key is the variable name string then indices are keys in another nested dictionary:
```python
 var_dict = {'temperature': 300,
             'flow_mass_phase_comp': {('Liq', 'NaCl'): 1,
                                      ('Liq', 'H2O'): 2},
            }
```
3. The equivalent form used in parameter sweep where the dictionary key is the variable itself:
```python
 var_dict = {blk.temperature: 300,
             blk.flow_mass_phase_comp['Liq', 'NaCl']: 1,
             blk.flow_mass_phase_comp['Liq', 'H2O']: 2,
            }
```
4. Then this suggestion where the dictionary key is the variable name and index in a single string:
```python
 var_dict = {'temperature': 300,
             'flow_mass_phase_comp[Liq, NaCl]': 1,
             'flow_mass_phase_comp[Liq, H2O]': 2,
            }
```
I'm not sure what other ways IDAES or ProteusLib has used to input variables and an attribute as an argument, but they might be floating out there. 

I liked method 1 better than method 2 because I think it is cleaner than the nested dictionary. Method 3 makes a lot of sense too, but sometimes these dictionaries need to be created before the model is made, like in the property test harness where the solution values need to be provided to the test which later creates the model. Method 4 is interesting, I had not used find_component or component before.

Reply:
So, API discussions are open-ended, and as noted there is not a standard defined for this yet.

Regarding Method 2 - this was used for the initialization as it is a very specific case and that structure fits well with the structure. I.e., in initialization we have a pre-defined list of state variables which we can iterate over, so it is just a matter of matching names from the list with the keys in the `state_args` dict.

Reply:
Also, I just added this topic to the IDAES Core Developer meeting agenda for next week.

Reply:
@adam-a-a and @bknueven which forms do you guys prefer and why? I'm leaning towards form 1 because it is at least consistent with the property test harness that I added to ProteusLib. I agree full consistency should be the goal, but I won't be able to do that in this PR.

Reply:
The `parameter_sweep` version (form 3) is subtly different -- the *keys* to the dictionary are just names, an the var and values are store in a value tuple, e.g.
```python
var_dict = { 'temp' : (blk.temperature, 300),
             'flow_mass_phase_comp_liq_NaCl' : (blk.flow_mass_phase_comp['Liq', 'NaCl'], 1),
             'flow_mass_phase_comp_liq_H2O': (blk.flow_mass_phase_comp['Liq', 'H2O'], 2),
            }
 ```
 
 In order to use python objects as "keys" you need to use pyomo.environ.ComponentMap, e.g.
 ```python
import pyomo.environ as pyo

var_map = ComponentMap()
var_map[blk.temperature] =  300
var_map[blk.flow_mass_phase_comp['Liq', 'NaCl']] = 1
var_map[blk.flow_mass_phase_comp['Liq', 'H2O']] = 2
 ```

Reply:
@bknueven yeah oops I should of been more clear that it wasn't exactly the form the parameter_sweep tool uses. I knew it used the short name, but I was trying to provide an option of a different key such as the actual variable itself.

I would actually prefer form 3 generally, but in a lot of cases these dictionaries are inputs into functions that will later create the model, so form 3 can't be used in those cases.

Reply:
Right -- I agree that the ComponentMap version is the best when the model already exists. I has lots of advantages, such as being able to write:
```python
for var, val in var_map.items():
    # for example
    var.fix(val)
```
which is much cleaner than any of the other alternatives. I'm not necessarily a big fan for user-facing interfaces you also have to *tell* the user to use `ComponentMap` in place of `dict`, which may be confusing for new python/Pyomo users.